### PR TITLE
Finetune task

### PR DESCRIPTION
**Pull request by dangkai4u**
此pr目的是实现微调下游任务，可用于评估预训练语言模型的能力。

目前实现了两个benchmark。clue用于中文模型，glue用于英文模型。
- [x] 通过prepare_data目录下脚本一键下载并提取数据。
- [x] 采用huggingface的方法，抽象微调阶段数据处理组建。InputExample存储单个原始样本（文本形式），InputFeatures存储单个样本的特征（input_ids，attention_mask，token_type_ids，label），DataProcessor针对每一个数据集读取并解析。
- [x] 实现了一些简单的评价指标（accuracy，f1，pearson，spearmanr），并标明每个数据集对应的评价指标。
- [x] datasets目录下，是微调阶段的数据集，调用get_processor和convert_examples_to_features方法，构造数据。

- [x] 数据构造时，采用的是BERT的构造方法，也就是[CLS] tokens_a [SEP] tokens_b [SEP]或[CLS] tokens_a [SEP]形式。gpt不能基于这个benchmark评估，可以不管。t5可以采用这个方法，但目前没有想好如何构造。
    - 这个问题，通过引入EncodePattern枚举变量实现，这个变量控制如何构造数据，目前支持了两种模式，更多模式看之后的需求。
- [x] 构造数据时默认构造了token_type_ids，但有的模型不需要，这点感觉也比较难以统一。
    - 默认构造，是否使用，看具体模型使用时的参数选择。

### CODE REVIEW

**Comment by thinksoso** [comment]
https://github.com/Oneflow-Inc/OneTeam/issues/1153 有一个bug，解决了这个应该就能合并了

**Comment by NotEmpty111** [comment]
> 
运行样例
bash tools/train.sh tools/train_net.py projects/text_classification/configs/config.py 2 train.train_iter=10

File: projects/text_classification/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Text classification projects
+
```
Comment (rentainhe) [inline projects/text_classification/README.md:2]: 在这里加一下中文README的链接把

| **English** | [中文简体]() |

**Comment by thinksoso** [comment]
运行:
bash tools/train.sh tools/train_net.py projects/text_classification/config.py 2 train.train_iter=10

File: projects/text_classification/README.md (hunk 1 above)
Comment (rentainhe) [reply projects/text_classification/README.md:2]: 中文README可以命名为`README_zh-CN.md`, 一般不会是`README_zh.md`

File: projects/text_classification/README_zh.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+在文本分类任务上微调预训练语言模型。[GLUE](https://gluebenchmark.com/) 是一个英文文本分类常见的基准任务，通常用它评估预训练语言模型的理解能力。
+
+GLUE 由9个子任务组成，包括cola、sst2、mrpc、stsb、qqp、mnli、qnli、rte、wnli。你可以执行通过命令实现一键下载并提取数据：
+```bash
+cd projects/text_classification
+python3 dataset/download_glue_data.py
+```
+
+这里示范如何微调该任务：
```
Comment (rentainhe) [inline projects/text_classification/README_zh.md:14]: 示范如何微调 -> `可以通过以下命令在此任务上进行finetune`

File: projects/text_classification/configs/config.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+from omegaconf import OmegaConf
+
+from configs.common.data.bert_dataset import tokenization
```
Comment (rentainhe) [inline projects/text_classification/configs/config.py:3]: 这里改成用`get_config`接口吧

```python
from libai.config import get_config

tokenization = get_config("common/data/bert_dataset.py")
```
下面的一些接口同理, 只需要本地安装后就可以这样import了, `pip install -e .`

File: projects/text_classification/configs/config.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            data_dir="/DATA/disk1/liuchi/work/clue_data/afqmc",
+            tokenizer=tokenization.tokenizer,
+            max_seq_length=512,
+            mode="dev",
+        ),
+        num_workers=4,
+    ),
+]
+
+model_cfg.update(
```
Comment (rentainhe) [inline projects/text_classification/configs/config.py:48]: 这里也可以直接在import的model的config上改对吧

File: projects/text_classification/dataset/clue_dataset.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+# limitations under the License.
+
+import logging
+import os
+import time
+from enum import Enum
+from typing import Optional, Union
+
+import oneflow as flow
+from filelock import FileLock
```
Comment (rentainhe) [inline projects/text_classification/dataset/clue_dataset.py:23]: filelock的import放到oneflow上面 typing下面吧

File: projects/text_classification/dataset/download_clue_data.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+""" Script for downloading all CLUE data.
```
Comment (rentainhe) [inline projects/text_classification/dataset/download_clue_data.py:1]: 诶这里需要加license嘛~

File: requirements.txt [hunk 7]
```diff
@@ -1,6 +1,8 @@
 boto3
 botocore
 cloudpickle
+flowvision
+filelock
 flowvision==0.1.0
```
Comment (rentainhe) [inline requirements.txt:6]: 这儿似乎已经有指定版本的flowvision了~

File: projects/text_classification/configs/config.py (hunk 4 above)
Comment (thinksoso) [reply projects/text_classification/configs/config.py:48]: 也可以，这样方便些吧

Comment (rentainhe) [reply projects/text_classification/configs/config.py:48]: 昂，如果你觉得这样语义清晰一些其实也OK我感觉

File: projects/text_classification/README_zh-CN.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# 文本分类任务
+[English](/projects/text_classification/README.md) | 简体中文
```
Comment (CPFLAME) [inline projects/text_classification/README_zh-CN.md:2]: 这里简体中文可以加粗,  **简体中文**

File: projects/text_classification/configs/config.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+from projects.text_classification.dataset import ClueDataset
+
+tokenization = get_config("common/data/bert_dataset.py").tokenization
+optim = get_config("common/optim.py").optim
+model_cfg = get_config("common/models/bert.py").cfg
+graph = get_config("common/models/graph.py").graph
+train = get_config("common/train.py").train
+
+tokenization.tokenizer = LazyCall(BertTokenizer)(
+    vocab_file="/DATA/disk1/liuchi/work/bert-base-chinese-vocab.txt",
```
Comment (CPFLAME) [inline projects/text_classification/configs/config.py:17]: 这里我觉得可以改成相对路径,   其实可以把这些所有需要用到的数据集, 上传到oss, 然后在程序里面用wget 下载下来.

可以参考https://github.com/Oneflow-Inc/libai/tree/main/tests/models 里面的做法

File: .flake8 [hunk 10]
```diff
@@ -15,4 +15,5 @@ per-file-ignores =
   tests/config/**.py:F401,E402
   tests/**.py: E402
   tools/**.py: E402
-  projects/**/configs/**.py:F401
\ No newline at end of file
+  projects/**/configs/**.py:F401
+  projects/text_classification/dataset/download_glue_data.py: E501
```
Comment (L1aoXingyu) [inline .flake8:19]: 这里为什么要对这个特定的文件增加一个规则，这种规则建议直接去对应的代码里面通过 #noqa 的方式去增加

File: requirements.txt [hunk 11]
```diff
@@ -1,6 +1,7 @@
 boto3
 botocore
 cloudpickle
+filelock
```
Comment (L1aoXingyu) [inline requirements.txt:4]: 和 project 依赖的 python 放到对应的 project 的 requirements.txt 里面

File: libai/evaluation/reg_evaluator.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+class RegEvaluator(DatasetEvaluator):
+    def __init__(self, cfg):
+        self.cfg = cfg
+        self._predictions = []
+
+    def reset(self):
+        self._predictions = []
+
+    def process(self, inputs, outputs):
+        # FIX ME: support dict args, not implement in graph right now
```
Comment (L1aoXingyu) [inline libai/evaluation/reg_evaluator.py:39]: 现在 graph 已经支持 dict args，可以改一下

File: libai/evaluation/reg_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+from scipy.stats import pearsonr, spearmanr
+
+from libai.utils import distributed as dist
+
+from .evaluator import DatasetEvaluator
+
+logger = logging.getLogger(__name__)
+
+
+class RegEvaluator(DatasetEvaluator):
```
Comment (L1aoXingyu) [inline libai/evaluation/reg_evaluator.py:30]: 这个 evaluator 为什么不放到 project 里面，有什么原因吗

File: projects/text_classification/modeling/model.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+import oneflow as flow
+from oneflow import nn
+
+from libai.layers import Linear
+from libai.models.bert_model import BertModel
+from libai.models.utils import init_method_normal
+from libai.utils import distributed as dist
+
+from .load_megatron_weight import load_megatron_bert
```
Comment (CPFLAME) [inline projects/text_classification/modeling/model.py:26]: 这行可以放到 48 ~ 49 行之间

File: libai/evaluation/reg_evaluator.py (hunk 13 above)
Comment (thinksoso) [reply libai/evaluation/reg_evaluator.py:30]: 这个evaluator在text_classifications里面没用到，我猜dangkai只是想拓展一下evaluation, 多提供几个指标？

File: projects/text_classification/configs/config.py (hunk 9 above)
Comment (thinksoso) [reply projects/text_classification/configs/config.py:17]: 这个bert-base-chinese.txt 还没有提供下载脚本，感觉不是很有必要专门为了某个特定的模型搞个下载脚本, 我理解projects更多是提供一个demo，方便修改。数据倒是可以默认一个存储路径，我改一下。

File: .flake8 (hunk 10 above)
Comment (thinksoso) [reply .flake8:19]: 我研究一下flake8~

**Comment by rentainhe** [comment]
需要执行一下`bash dev/linter.sh`修复一下格式问题~

**Review by L1aoXingyu — APPROVED** [review]

**Comment by rentainhe** [comment]
> 需要执行一下`bash dev/linter.sh`修复一下格式问题~

还有unittest里缺少了一个module的安装，可能需要在requirements哪儿再改一下~

**Review by rentainhe — APPROVED** [review]
